File: Anul II/Medii de proiectare si programare/Laborator/lab06/networking/src/main/java/com/ubb/protocol/ServicesProxy.java
package com.ubb.protocol;

import com.ubb.IMainObserver;
import com.ubb.IContestServices;
import com.ubb.exceptions.ContestDataException;
import com.ubb.model.*;
import com.ubb.model.data.RaceDTO;
import com.ubb.protocol.request.*;
import com.ubb.protocol.response.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ServicesProxy implements IContestServices {
    private final String host;
    private final int port;
    private IMainObserver client;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;

    private final BlockingQueue<Response> responseQueue;
    private volatile boolean finished;

    public ServicesProxy(String host, int port) {
        this.host = host;
        this.port = port;
        responseQueue = new LinkedBlockingQueue<>();
    }

    @Override
    public void login(String username, String password, IMainObserver client) throws ContestDataException {
        initializeConnection();
        sendRequest(new LoginRequest(username, password));
        Response response = readResponse();
        if (response instanceof OKResponse) {
            this.client = client;
            return;
        }
        if (response instanceof ErrorResponse) {
            closeConnection();
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
    }

    @Override
    public void logout(User user, IMainObserver client) throws ContestDataException {
        sendRequest(new LogoutRequest(user));
        Response response=readResponse();
        closeConnection();
        if (response instanceof ErrorResponse){
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
    }

    @Override
    public Collection<Participant> getParticipantsByTeam(Long teamID) throws ContestDataException {
        sendRequest(new GetParticipantsByTeamRequest(teamID));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        GetParticipantsByTeamResponse participantsResponse = (GetParticipantsByTeamResponse) response;
        return participantsResponse.getTeamParticipants();
    }

    @Override
    public Optional<Participant> getParticipantByData(Participant participant) throws ContestDataException {
        sendRequest(new GetParticipantByDataRequest(participant));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        ParticipantResponse participantResponse = (ParticipantResponse) response;
        return participantResponse.getParticipant();
    }

    @Override
    public Optional<Race> getRaceByName(String name) throws ContestDataException {
        sendRequest(new GetRaceByNameRequest(name));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        RaceByNameResponse raceResponse = (RaceByNameResponse) response;
        return raceResponse.getRace();
    }

    @Override
    public Collection<RaceEntry> getEntriesByRace(Long raceID) throws ContestDataException {
        sendRequest(new GetEntriesByRaceRequest(raceID));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        EntriesByRaceResponse entriesResponse = (EntriesByRaceResponse) response;
        return entriesResponse.getEntries();
    }

    @Override
    public Collection<Race> getRacesByEngineCapacity(Integer engineCapacity) throws ContestDataException {
        sendRequest(new GetRacesByEngineCapacityRequest(engineCapacity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        RacesByCapacityResponse racesResponse = (RacesByCapacityResponse) response;
        return racesResponse.getRaces();
    }

    @Override
    public Collection<RaceDTO> getRacesWithParticipantCount() throws ContestDataException {
        sendRequest(new GetRacesWithParticipantCountRequest());
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        RacesWithParticipantsResponse racesResponse = (RacesWithParticipantsResponse) response;
        return racesResponse.getRaces();
    }

    @Override
    public void saveRaceEntries(List<RaceEntry> newEntities) throws ContestDataException {
        sendRequest(new CreateRaceEntriesRequest(newEntities));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
    }

    @Override
    public Collection<Race> getRacesWhereNotRegisteredAndEngineCapacity(Long participantID, Integer engineCapacity) throws ContestDataException {
        sendRequest(new GetUnregisteredRacesRequest(participantID, engineCapacity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        RacesByCapacityResponse racesResponse = (RacesByCapacityResponse) response;
        return racesResponse.getRaces();
    }

    @Override
    public Optional<Team> getTeamByName(String teamName) throws ContestDataException {
        sendRequest(new GetTeamByNameRequest(teamName));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
        TeamResponse teamResponse = (TeamResponse) response;
        return teamResponse.getTeam();
    }

    @Override
    public void saveParticipant(Participant newEntity) throws ContestDataException {
        sendRequest(new CreateParticipantRequest(newEntity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
    }

    @Override
    public void saveRace(Race newEntity) throws ContestDataException {
        sendRequest(new CreateRaceRequest(newEntity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getMessage());
        }
    }

    private void initializeConnection() throws ContestDataException {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startResponseReader();
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void startResponseReader() {
        Thread responseReader = new Thread(new ReaderThread());
        responseReader.start();
    }

    private void handleUpdate(UpdateResponse update) {
        try {
            if (update instanceof final NewParticipantResponse participantResponse) {
                client.participantAdded(participantResponse.getParticipant());
            }

            if (update instanceof final NewRaceResponse raceResponse) {
                client.raceAdded(raceResponse.getRace());
            }

            if (update instanceof final UpdatedRacesResponse racesResponse) {
                client.raceEntriesAdded(racesResponse.getRaces());
            }
        } catch (ContestDataException e) {
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) throws ContestDataException {
        try {
            output.writeObject(request);
            output.flush();
        } catch (IOException ioException) {
            throw new ContestDataException("Error sending object: " + ioException);
        }
    }

    private Response readResponse() throws ContestDataException {
        Response response = null;
        try {
            response = responseQueue.take();
        } catch (InterruptedException interruptedException) {
            interruptedException.printStackTrace();
        }
        return response;
    }

    private class ReaderThread implements Runnable{
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    System.out.println("response received " + response);
                    if (response instanceof UpdateResponse){
                        handleUpdate((UpdateResponse) response);
                    } else {
                        try {
                            responseQueue.put((Response)response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException | ClassNotFoundException e) {
                    System.out.println("Reading error "+e);
                }
            }
        }
    }

}

File: Anul II/Medii de proiectare si programare/Laborator/lab07/Ubb.BikeContest.Networking/ServicesProxy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;
using Ubb.BikeContest.Model;
using Ubb.BikeContest.Model.data;
using Ubb.BikeContest.Networking.Request;
using Ubb.BikeContest.Networking.Request.Request;
using Ubb.BikeContest.Networking.Response;
using Ubb.BikeContest.Networking.Response.Response;
using Ubb.BikeContest.Services;
using Ubb.BikeContest.Services.src.Exceptions;

namespace Ubb.BikeContest.Networking
{
    public class ServicesProxy : IContestServices
    {
        private readonly string host;
        private readonly int port;
        private IMainObserver client;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;
        private readonly Queue<IResponse> responses;
        private volatile bool finished;
        private EventWaitHandle _waitHandle;

        public ServicesProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<IResponse>();
            InitialiseConnection();
        }

        public virtual User Login(string username, string passwordToken, IMainObserver client)
        {
            SendRequest(new LoginRequest(username, passwordToken));
            IResponse response = ReadResponse();
            if (response is OkResponse)
            {
                this.client = client;
                return ((OkResponse)response).User;
            }
            if (response is ErrorResponse)
            {
                ErrorResponse error = (ErrorResponse)response;
                CloseConnection();
                throw new ContestDataException(error.Message);
            }
            return null;
        }

        public void Logout(User user, IMainObserver client)
        {
            SendRequest(new LogoutRequest(user));
            IResponse response = ReadResponse();
            CloseConnection();
            if (response is ErrorResponse)
            {
                ErrorResponse error = (ErrorResponse)response;
                throw new ContestDataException(error.Message);
            }
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
        }

        private void SendRequest(IRequest request)
        {
            try
            {
                formatter.Serialize(stream, request);
                stream.Flush();
            }
            catch (Exception exception)
            {
                throw new ContestDataException("Error sending object " + exception);
            }
        }

        private IResponse ReadResponse()
        {
            IResponse response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
            return response;
        }

        private void InitialiseConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
        }

        private void StartReader()
        {
            Thread threadWorker = new Thread(Run);
            threadWorker.Start();
        }

        private void HandleUpdate(UpdateResponse update)
        {
            if (update is NewParticipantResponse)
            {
                NewParticipantResponse response = (NewParticipantResponse)update;
                Console.WriteLine("New participant added: " + response.Participant);
                try
                {
                    client.ParticipantAdded(response.Participant);
                }
                catch (ContestDataException exception)
                {
                    Console.WriteLine(exception.StackTrace);
                }
            }

            if (update is UpdatedRacesResponse)
            {
                UpdatedRacesResponse response = (UpdatedRacesResponse)update;
                Console.WriteLine("Updated races");
                try
                {
                    client.RaceEntriesAdded(response.Races);
                }
                catch (ContestDataException exception)
                {
                    Console.WriteLine(exception.StackTrace);
                }
            }
        }

        public virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    object response = formatter.Deserialize(stream);
                    Console.WriteLine("Response received: " + response);
                    if (response is UpdateResponse)
                    {
                        HandleUpdate((UpdateResponse)response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue((IResponse)response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception exception)
                {
                    Console.WriteLine("Reading error: " + exception);
                }
            }
        }

        public IEnumerable<RaceDto> GetRacesWithParticipantCount()
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Team> FindAllTeams()
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Participant> GetParticipantsByTeam(long id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Participant> FindAllParticipants()
        {
            throw new NotImplementedException();
        }
    }
}
